chichiwang/exercise-js
# Instrument React Router into the project

## Description
Linked to Issue: #19
Instrument a Router with a single route `/` into the project.

## Changes
* Add devDependencies to `package.json`:
  * [@testing-library/react](https://testing-library.com/docs/react-testing-library/intro/) - unit testing utilities for React
  * [@testing-library/jest-dom](https://testing-library.com/docs/ecosystem-jest-dom/) - dom element matchers for React component unit testing
* Add dependency to `package.json`:
  * [react-router-dom](https://reactrouter.com/en/main)
* Change filename `app/index.tsx` to `app/App.tsx` in preparation for convention changes
  * File names should match the component name
* Wrap `App` component in [React.StrictMode](https://react.dev/reference/react/StrictMode) component
* Create `app/AppRoutes.tsx` component using react-router
  * Create a `/` root route to render the `Home` component
* Update `client/entry.tsx` to wrap the `App` in a `BrowserRouter` for client-side hydration
* Update `server/index.tsx` to defer routing to the React `App`, wrapped in a `StaticRouter` for server-side rendering
* Configure `jest.config.json` to set `testEnvironment` to `jsdom` for component test cases

## Steps to QA
* Pull down and switch to this branch
* Run `npm run build:dev` and ensure the build runs without issue
  * Run `node dist/index.js` to ensure the server runs without issue. Verify in browser.
* Run `npm start` to ensure watcher builds and runs the server without issue
  * Verify in browser
